src/mcts/search.cc

```diff
@@ -1291,9 +1307,38 @@ void SearchWorker::ExtendNode(Node* node) {
       return;
     }
 
-    if (history_.Last().GetRepetitions() >= 2) {
+    const auto repetitions = history_.Last().GetRepetitions();
+    const auto twofolddrawlevel = params_.GetTwoFoldDrawLevel();
+    bool validtwofold = false;
+    // Mark two-fold repetitions as draws according to settings
+    // Depth starts with 1 at root, so number of plies in PV is depth - 1
+    if (repetitions >= 2) {
       node->MakeTerminal(GameResult::DRAW);
       return;
+    } else if (repetitions == 1 && twofolddrawlevel > 0) {
+      if (twofolddrawlevel == 3) {
+        // always mark as draw
+        validtwofold = true;
+      } else if (twofolddrawlevel == 2 && depth - 1 >= 4) {
+        // only mark as draw if depth of extended node is >= 4
+        validtwofold = true;
+      } else if (twofolddrawlevel == 1 && depth - 1 >= 4 && depth - 1 >=
+                 history_.ComputePliesSinceFirstRepetition()) {
+        // check whether first repetition happened at root or in the tree
+        // don't mark as draw if repetition happened in the game history
+        validtwofold = true;
+      }
+    }
+    // if node turned out to be a valid twofold, mark it as Terminal::Twofold
+    if (validtwofold) {
+      const auto cyclelength = history_.ComputePliesSinceFirstRepetition();
+      // logging for debugging purpose
+      LOGFILE << "== marked twofold draw == depth: " << depth - 1;
+      LOGFILE << "== plies since first repetition: " << cyclelength;
+      // use plies since first repetition as moves left; exact if forced draw.
+      node->MakeTerminal(GameResult::DRAW, (float)cyclelength,
```

Comment:
Clever stashing the value to avoid needing the position later. Will this have strange moves-left-effect behavior? Does this value want to be doubled assuming the best move to play from a 2-fold is to repeat the cycle into 3-fold? Otherwise a winning position might try to give a boost to this shorter-looking move?

Also, I don't think the cast is necessary.

Reply:
If I understand your concern correctly: It actually is the correct `m_` assuming that repetition of the move is forced from here. The 2-fold is set as a terminal with `m_=cyclelength` (instead of 0 as other terminals). Also, MLH with Q scaling doesn't calculate any bonus for draws, unless we use `drawscore`.